### ISSUE TITLE
Profile Mastered Users Can't be updated via Terraform User Updates

### ISSUE THREAD

**Issue by virgofx** [okta/terraform-provider-okta#1569]
When an Okta profile is mastered via a separate app (e.g. Workday, BambooHR), the Terraform provider (and thus public API) is unable to update certain base attribute user fields that have been set to be OKTA mastered.

---

For example, the following profile is mastered via BambooHR:
<p align="center">
   <img src="https://github.com/okta/terraform-provider-okta/assets/739719/2b346ddf-9584-46e2-ba7b-90f4f27dcb9e" />
</p>

One of our custom Okta fields, in this case `organization`, is set to be Okta mastered as shown below:
<p align="center">
  <img src="https://github.com/okta/terraform-provider-okta/assets/739719/2ea2ade1-3306-4598-a8cb-756d6aa3a973" />
</p>

Upon clicking the **Edit** button in the Okta Admin UI edit profile page, we can clearly see the fields that can be modified via Okta as these will be editable (have input boxes). Observe in the following screenshot, we can correctly change `organization` but not the other Bamboo mastered fields - `division`, `employeeNumber`, `department`.

<p align="center">
  <img src="https://github.com/okta/terraform-provider-okta/assets/739719/7a1126fd-545b-483d-88fa-468b6ff8cb46" />
</p>

Let's change the organization, via the UI.  Organization changes from `ChangeMe` to `New Org`. This hits the admin web API endpoint, which is obviously different; however, the request succeeds and the user's organization is updated.

```
Request URL: https://XXXXX-admin.okta.com/api/v1/users/00u4kh9lncTpBIBus697
Request Method: PUT
Status Code: 200
Payload: {..., profile: {..., organization: "New Org", ...}, ...}
```

The user's `organization` field is successfully updated.

<p align="center">
  <img src="https://github.com/okta/terraform-provider-okta/assets/739719/5753f7ed-e167-4e46-b7a8-f62278790a29" />
</p>

Let's attempt to update this user which was previously created in Terraform, subsequently imported and partially managed via BambooHR.

```
resource "okta_user" "dot_testing" {
  login              = "dot.testing@XXXXX.com"
  email              = "dot.testing@XXXXX.com"
  first_name         = "Dot"
  last_name          = "Testing"

  # Okta mastered fields
  organization       = "New New Org"

  lifecycle {
    ignore_changes = [
      city,            # Confirmed BambooHR managed and mapped
      department,      # Confirmed BambooHR managed and mapped
      display_name,    # Confirmed BambooHR managed and mapped
      division,        # Confirmed BambooHR managed and mapped
      employee_number, # Confirmed BambooHR managed and mapped
      first_name,      # Confirmed BambooHR managed and mapped
      last_name,       # Confirmed BambooHR managed and mapped
      manager,         # Confirmed BambooHR managed and mapped
      manager_id,      # Confirmed BambooHR managed and mapped
      middle_name,     # Confirmed BambooHR managed and mapped
      mobile_phone,    # Confirmed BambooHR managed and mapped
      nick_name,       # Confirmed BambooHR managed and mapped
      postal_address,  # Confirmed BambooHR managed and mapped
      primary_phone,   # Confirmed BambooHR managed and mapped
      profile_url,     # Confirmed BambooHR managed and mapped
      second_email,    # Confirmed BambooHR managed and mapped
      state,           # Confirmed BambooHR managed and mapped
      street_address,  # Confirmed BambooHR managed and mapped
      title,           # Confirmed BambooHR managed and mapped
      zip_code,        # Confirmed BambooHR managed and mapped
    ]
  }
```

Terraform hits the public API endpoint for Users update and won't allow any changes. An update at this point to change the `organization`  seems reasonable as shown in the Terraform apply confirmation.

```
# okta_user.dot_testing will be updated in-place
  ~ resource "okta_user" "dot_testing" {
        id                                  = "00u4kh9XXXXXXXXXXXXXX"
      ~ organization                        = "New Org" -> "New New Org"
        # (28 unchanged attributes hidden)
    }
```

However, upon applying, Terraform **errors** out as such:

```
 Error: failed to update user: the API returned an error: Operation failed because user profile is mastered under another system
```

It would be great to get some insight here, potentially get the public API updated to support this. Not sure if this is the best spot for upstream bugs; however, since this can be fairly easily reproduced I figured we could capture this issue here and additionally add an API/internal ticket that references this.

### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

1.3.5 / Okta: 4.0.0

### Affected Resource(s)

- `okta_user`

**Comment by duytiennguyen-okta** [comment]
@virgofx we need to investigate further. We also wondering if we should do a Patch vs Put. The outstanding question is the monolith able to update profiles of third party app but that behavior is not exposed through public API

**Comment by virgofx** [comment]
@duytiennguyen-okta Thanks for the update. Since we just moved to a system where we have third party profiles now, we're unable to leverage the API anymore (and thus Terraform for our users) which is a huge blocker. In fact, our pipeline is now broken as we've had a git-ops centered approach to Okta for while. Thus, we are huge proponents of allowing the API to update the `OKTA` mastered fields of 3rd party mastered profiles -- similar to how the Okta UI does. Is there anything I can do to help advance this case forward internally?

**Comment by monde** [comment]
@virgofx can you run tf with `TF_LOG=debug` and tell me if the provider is hitting the API with the same URI and http method? When you were debugging the Admin UI you found it to be doing:

```
Request URL: https://XXXXX-admin.okta.com/api/v1/users/00u4kh9lncTpBIBus697
Request Method: PUT
Status Code: 200
Payload: {..., profile: {..., organization: "New Org", ...}, ...}
```

I wonder if the TF provider will be hitting the non-admin domain `XXXXX.okta.com` but the same API URI `/api/v1/users/00u4kh9lncTpBIBus697` and method `PUT`:

```
Request URL: https://XXXXX.okta.com/api/v1/users/00u4kh9lncTpBIBus697
Request Method: PUT
Status Code: TBD
Payload: {..., profile: {..., organization: "New Org", ...}, ...}
```

In my dealings with the Okta monolith I've not experienced org domain vs admin org domain being something that is checked on the HTTP requests. e.g. the public API endpoints `/api/v1/` are all that matters. But that doesn't mean that it doesn't occur, I just may not be aware of that as a check.

Interested in seeing if the provider is doing a `PUT` on `/api/v1/users/00u4kh9lncTpBIBus697`.

Let me know what you see, it will help me present the correct question to the feature team or find the answer on my own.

**Comment by virgofx** [comment]
Confirmed as you suspected that the Terraform Provider (and thus the core Okta Goland SDK) is hitting the main API endpoint (which is what I'd expect). So, it's more of a what is happening in the Admin UI API endpoint that is unique to allow this and can we just copy that similar logic over to the main endpoint? If so, likely no changes in Terraform land.

```
---[ REQUEST ]---------------------------------------
PUT /api/v1/users/00uif5ibm63HfENpw696 HTTP/1.1
Host: XXXXXX.okta.com
User-Agent: okta-sdk-golang/2.17.0 golang/go1.20.5 windows/amd64 okta-terraform/4.1.0
Content-Length: 1099
Accept: application/json
Authorization: SSWS *****************************
Content-Type: application/json
Accept-Encoding: gzip
```

```
---[ RESPONSE ]--------------------------------------
HTTP/2.0 403 Forbidden
...
{
 "errorCode": "E0000023",
 "errorSummary": "Operation failed because user profile is mastered under another system",
 "errorLink": "E0000023",
 "errorId": "oaeS55cRywVR8CEQadoEYNuNw",
 "errorCauses": []
}
```

**Comment by virgofx** [comment]
Update:

It appears that both the main API user endpoint from the admin-ui and the standard API endpoint, both can successfully update the non-Okta mastered attributes, if and only if:

- All of the `profile` attributes are specified. 
- Specifically, fields that are mastered and are strings and contain "" must be sent as an empty string instead of being omitted (I believe this is the key issue. Terraform here is dropping some non-Okta mastered fields since the returned value was `""`)

If one specifies the same (unaltered) values for non-Okta mastered fields into the main API, it appears one can successfully update the Okta mastered fields. 

---

So, after investigation, it appears this can be fixed by passing in empty strings for values that are Bamboo mastered. Let's take an example in this case.

- `profile_url` is non-Okta mastered. 
- BambooHR sets and maps this value over as an empty string or even setting it null
- Updates via API will work if the `profile_url` is specified as `""` (An empty string). Specifically, the JSON includes a `profile` attribute `profile_url = ""`. 
- The issue with Terraform is that all of the profile attributes are marked as optional. When populating the JSON object for the update calls it uses the `d.GetOk()` function which only returns the value when it's not null or not an empty string. https://github.com/okta/terraform-provider-okta/blob/master/okta/user.go#L229   
- If the above line was strict and always setting an empty string it would fix the issue. 

So, this kind of spurs a secondary issue. Okta has the concept of allowing some properties to be `null`; however, it appears in the UI it does change nulls to blank strings `""` on any first update. But again, via the API, one can omit properties for an Okta mastered API call, which effectively nulls the value, or alternatively using a mapping for a non-okta Mastered attribute and setting it to `null` has the same effect.

----
TLDR;

- The Terraform user update calls will **drop (not send)** any attribute that has an intrinsic `null` value or empty string `""`.
- For users that are non-Okta mastered, all attributes must match exactly. This includes effective `null` to `null` matching and string to string `""` to `""`.
- Terraform user updates will pass the previous stored state when an attribute is lifecycle_ignored. It appears that there are a few attributes that require having the current value of the non-Okta mastered attribute to ensure the API call goes through. (Specifically, `profile_url` must match **exactly**, whereas `first_name`, `display_name`, etc. do not need to match ) This means, one could theoretically enforce hybrid terraform-IAC with Bamboo mastered attributes, by plumming the output from `data "okta_user" ...` calls back into the user and then removing the `lifecycle_ignored` attribute for non-Okta mastered attributes. Unfortunately, this means, 



---

Since we have a workaround now, I'm going to close the issue.

For those that may experience this, instead of excluding non-Okta mastered fields via lifecycle rules, one instead needs to specify them directly from output from a `data "okta_user"` resource to ensure equality.
